integration/bootnodes_test.go

```diff
@@ -0,0 +1,137 @@
+package integration
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// startV4 starts an ephemeral discovery V4 node.
+func startV4(t *testing.T) *discover.UDPv4 {
+	socket, ln, cfg, err := listen()
+	if err != nil {
+		t.Fatal(err)
+	}
+	disc, err := discover.ListenV4(socket, ln, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return disc
+}
+
+func listen() (*net.UDPConn, *enode.LocalNode, discover.Config, error) {
+	var cfg discover.Config
+	cfg.PrivateKey, _ = crypto.GenerateKey()
+	db, _ := enode.OpenDB("")
+	ln := enode.NewLocalNode(db, cfg.PrivateKey)
+
+	socket, err := net.ListenUDP("udp4", &net.UDPAddr{IP: net.IP{0, 0, 0, 0}})
+	if err != nil {
+		db.Close()
+		return nil, nil, cfg, err
+	}
+	addr := socket.LocalAddr().(*net.UDPAddr)
+	ln.SetFallbackIP(net.IP{127, 0, 0, 1})
+	ln.SetFallbackUDP(addr.Port)
+	return socket, ln, cfg, nil
+}
+
+func testBootnodes(t *testing.T, nodes []string, minPassRate float64, maxTrials int) {
+	if maxTrials == 0 {
+		t.Skip("trials disabled")
+	}
+
+	total := len(nodes)
+	failed := 0
+	minPassN := float64(total) * minPassRate // Minimum number of nodes that must be reachable for the test not to fail.
+
+	// Case where pass rate is epsilon (but non-zero) and rounding causes n nodes == 0; infer that at least just 1 node must pass.
+	if minPassN == 0 && minPassRate > 0 {
+		minPassN = 1
+	}
+
+	disc := startV4(t)
+
+nodesloop:
```

Comment:
I hate labels. I hate "continue nastyLabel" statements. Simple refactor - extract function that executes loop from lines 68-75, but instead of log+continue, just returns RTT and error. And log/count failures depending on return value.
```
unc checkENode(disc *discover.UDPv4, en *enode.Node, maxTrials int) (time.Duration, error) {
       var err error
       for i := 1; i <= maxTrials; i++ {
               start := time.Now()
               err = disc.Ping(en)
               if err == nil {
                       return time.Since(start), nil
               }
       }
       return 0, err
}
```

Reply:
https://github.com/etclabscore/core-geth/pull/26/commits/a542156c9d7c37cda7d240c901fbcadff11ffae3